pkg/apps/strategy/support/hooks.go

```diff
@@ -0,0 +1,198 @@
+	// can either be 'success' in case the hook finished successfully or the status message from the failed hook pod in case the hook
+	// failed to run. The error will be reported regardless of failure policy.
+	// This annotation should be used in CI systems or tests so they don't have to rely on deployer pod logs.
+	DeploymentHookResultAnnotation = "openshift.io/deployment-hook-result"
+
+	// DeploymentHookResultSuccess indicates the lifecycle hook succeeded without error.
+	DeploymentHookResultSuccess = "success"
+)
+
+type HookExecutor interface {
+	RunHook(hook *appsapi.LifecycleHook, controller *corev1.ReplicationController, stage string, timeout time.Duration) error
+}
+
+// LifecycleHookController operates the execution and lifecycle of deployment lifecycle hooks.
+type LifecycleHookController struct {
+	podClient   coreclient.PodsGetter
+	kubeClient  kubernetes.Interface
+	imageClient imageclientv1.Interface
+
+	recorder record.EventRecorder
+	out      io.Writer
+}
+
+var (
+	HookPodTimeoutError     error
+	DeployerPodTimeoutError error
+)
+
+func NewLifecycleHookController(kubeclientset kubernetes.Interface, imageClient imageclientv1.Interface, out io.Writer) *LifecycleHookController {
+	c := &LifecycleHookController{
+		kubeClient:  kubeclientset,
+		imageClient: imageClient,
+		out:         out,
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	if glog.V(5) {
+		eventBroadcaster.StartLogging(glog.Infof)
+	}
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeclientset.CoreV1().Events("")})
+
+	c.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "deployer"})
+
+	return c
+}
+
+func (c *LifecycleHookController) RunHook(hook *appsapi.LifecycleHook, rc *corev1.ReplicationController, hookType string, timeout time.Duration) error {
+	if len(hook.TagImages) > 0 {
+		h := &lifecycleTagHook{
+			imageClient: c.imageClient,
+			hookType:    hookType,
+			out:         c.out,
+		}
+		return h.run(hook, rc)
+	}
+
+	h := &lifecyclePodHook{
+		rcName:                rc.Name,
+		rcNamespace:           rc.Namespace,
+		podClient:             c.kubeClient.CoreV1().Pods(rc.Namespace),
+		out:                   c.out,
+		hookType:              hookType,
+		deployerStartTimeChan: make(chan time.Time),
+		hookStartTimeChan:     make(chan time.Time),
+		hookCompleteChan:      make(chan struct{}),
+		logStreamCompleteChan: make(chan struct{}),
+	}
+	// Start new filtered informer for all deployer managed pods
+	informer := coreinformers.NewFilteredPodInformer(c.kubeClient, rc.Namespace, 60*time.Second, cache.Indexers{}, func(options *metav1.ListOptions) {
+		options.LabelSelector = labels.SelectorFromSet(map[string]string{appsapi.DeployerPodForDeploymentLabel: rc.Name}).String()
+	})
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    h.handle,
+		UpdateFunc: h.handleUpdate,
```

Comment:
Don't you care about deletion?

Reply:
I'd argue about it, what if other controller mistakenly removes it, or even if user will do so, we should be self-healing.